llvm/lib/Support/SHA1.cpp

```diff
@@ -307,3 +307,39 @@ std::array<uint8_t, 20> SHA1::hash(ArrayRef<uint8_t> Data) {
   memcpy(Arr.data(), S.data(), S.size());
   return Arr;
 }
+
+std::string SHA1::hashToString(ArrayRef<uint8_t> Data, StringRef Mapping) {
+  constexpr unsigned BitsPerChar = 6;
+  constexpr unsigned CombinationPerChar = 1 << BitsPerChar;
+  assert(Mapping.size() >= CombinationPerChar);
+  std::array<uint8_t, 20> Hash = SHA1::hash(Data);
+  std::string Res;
+
+  unsigned Bits = 0;
+  unsigned Elem = 0;
+  constexpr uint8_t Mask1[] = {0b11111100, 0b00111111, 0b00001111, 0b00000011};
+  constexpr uint8_t Mask2[] = {0b00000000, 0b00000000, 0b11000000, 0b11110000};
+  constexpr uint8_t RShift1[] = {2, 0, 0, 0};
+  constexpr uint8_t LShift1[] = {0, 0, 2, 4};
+  constexpr uint8_t RShift2[] = {0, 0, 6, 4};
+  while (Elem < Hash.size()) {
+    uint8_t Pos = Bits >> 1;
+    assert(((Mask1[Pos] >> RShift1[Pos]) << LShift1[Pos] |
+            (Mask2[Pos]) >> RShift2[Pos]) == 0b00111111);
+    assert(((Mask1[Pos] >> RShift1[Pos]) << LShift1[Pos] &
+            (Mask2[Pos]) >> RShift2[Pos]) == 0);
+    uint8_t Value1 = ((Hash[Elem] & Mask1[Pos]) >> RShift1[Pos])
+                     << LShift1[Pos];
+    uint8_t Value2 = 0;
+    if (LLVM_LIKELY(Elem + 1 < Hash.size()))
+      Value2 = (Hash[Elem + 1] & Mask2[Pos]) >> RShift2[Pos];
+    Res.push_back(Mapping[Value1 | Value2]);
+    Bits += BitsPerChar;
+    if (Bits >= 8) {
+      Elem++;
+      Bits -= 8;
+    }
```

Comment:
Good obfuscation exercise!
Explain the FSM.
Not clear that at the end all the bits are consumed...
It would be simpler with a `std::range` of bits... :-)
Otherwise I wonder it would not be more understandable with a https://en.cppreference.com/w/cpp/utility/bitset `std::bitset<16>` that you fill by the left with `... | std::bitset<16> { c } << 8` when 1 char has been consumed and you empty it to the right with a `(... & std::bitset<16> { (1 << 6) - 1 }).to_ulong()`, then `>>= 6` and loop.

Reply:
> Good obfuscation exercise!
this wasn't the intent.

> Explain the FSM.
> Not clear that at the end all the bits are consumed...

I added assert at the end of the loop for that purpose.
> It would be simpler with a `std::range` of bits... :-)

this part of the code is build in c++14 mode so no std::range
and i would need index any way. to access the next element.
> Otherwise I wonder it would not be more understandable with a https://en.cppreference.com/w/cpp/utility/bitset `std::bitset<16>` that you fill by the left with `... | std::bitset<16> { c } << 8` when 1 char has been consumed and you empty it to the right with a `(... & std::bitset<16> { (1 << 6) - 1 }).to_ulong()`, then `>>= 6` and loop.

Reply:
These asserts do not explain anything.
After sleeping over this, I think that just having a `std::uint16_t` to play the role of accumulator/extractor and just using shift and mask to fill 8 bits and extract 6 bits are enough. No need for `std::bitset`.
This should be quite simpler and more understandable than the FSM approach.
Yes, I know that Xilinx is a hardware company, but writing FPGA RTL-like C++ code might be a little bit excessive. :-)

Reply:
using uint16_t greatly improved readability